connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -115,6 +115,49 @@ class SparkConnectPlanner(
   private lazy val pythonExec =
     sys.env.getOrElse("PYSPARK_PYTHON", sys.env.getOrElse("PYSPARK_DRIVER_PYTHON", "python3"))
 
+  // Some relation transform need to create Dataset, then get the logical plan from the Dataset.
+  // This method used to reuse the Dataset instead to discard it.
+  def transformRelationAsDataset(rel: proto.Relation): Dataset[Row] = {
```

Comment:
TBH I want to move away from constructing Datasets wholesale. In many cases there is no real need, and it is also expensive to do.

Reply:
Is this something you want to work on?

Reply:
> TBH I want to move away from constructing Datasets wholesale. In many cases there is no real need, and it is also expensive to do.

Yes. the current implementation create man duplicate datasets. We should reuse these datasets and reduce the overhead.

Reply:
> Is this something you want to work on?

Yes.